README.md
# PeaceL

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-04-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-04-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/todo-list/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/todo-list/compare/0.1.0...0.1.1


src/todo_list/mydb.clj
(ns todo_list.mydb 
  (:require    [datomic.client.api :as d]
))



;; Connection to the database hello

(def db-uri (d/client {:server-type :local 
                       :endpoint   "datomic:dev://localhost:4334/hello"}))

(d/create-database db-uri {:db-name "hello"})

(def conn (d/connect db-uri))







;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; (def cfg {:server-type :peer-server                ;;
;;           :access-key  "myaccesskey"               ;;
;;           :secret      "mysecret"                  ;;
;;           :endpoint    "localhost:4334"})          ;;
;;                                                    ;;
;; (def client (d/client cfg))                        ;;
;;                                                    ;;
;;                                                    ;;
;; (def conn (d/connect client {:db-name "hello"}))   ;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/todo_list/handlers.clj
(ns todo-list.handlers
  (:use  [hiccup.core]
         [hiccup.page]))


(defn the-handler                                                                                                         
 "When you send a request to the webapp, the ring adaptor (-main) converts the request to a map and sends it to the specified handler.This the handler function that takes the request map as its argument and returns a response map. " 
   [request]                                                                                                           (html [:h1 " Hello Good Human"]
          [:ul [:li "Huslte in the rain like no one is watching"]                       
                     [:li "Get this bread no matter what"]]))                                                                                                    

(defn goodbye
   "A route for goodbye"
   [request]
   (html [:h1 "Thank you so much for your visit."]
         [:p "We are very grateful of your presence!!"] ))


(defn about 
  "It gives you info about the prog."
  [request]
  {:status 200
   :body "<h1>Nkosinathi's simple web service, not bad...</h1>"
   :headers {}})

(defn req-info
  "Gives you information about the request that has been sent"
  [request]
  
  {:status 200
   :body (pr-str request)
   :headers {}})

(defn hello
  "Takes in a name from the request and display a personalised message"
  [request]
  (let [name (get-in request [:route-params :name])]
    {:status 200
     :body (str "Hello, "name ". How are you today ?")
     :headers {}}))

(def operators {"+" + "-" - ":" / "*" *})

(defn calculator
  "Gives a calculated result from the request's parameters"
  [request]
 (let [a  (Integer. (get-in request [:route-params :a]))
       b  (Integer. (get-in request [:route-params :b]))
       op (get-in request [:route-params :op])  
       the_op (get operators op)]
      
      
      
  (if the_op
    {:status 200
     :body  (str "Your result: " (the_op a b)) 
     :headers {}}
    {:status 404
     :body "I am not farmiliar with that operand. Try +,-,* or /"
     :headers {}})))


src/todo_list/core.clj
(ns todo-list.core
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]] ;push all our code changes to the application each time we save
            [compojure.core :refer [defroutes GET]]
            [compojure.route :refer [not-found]]
            [hiccup.core :refer :all]
            [hiccup.page :refer :all]
            [todo_list.mydb :refer :all]
))


(defn the-handler                                                                                                         
 "When you send a request to the webapp, the ring adaptor (-main) converts the request to a map and sends it to the specified handler.This the handler functyion that takes the request map as its argument and returns a response map. " 
   [request]                                                                                                           (html [:h1 " Hello Good Human"]
          [:ul [:li "Huslte in the rain like no one is watching"]                       
                     [:li "Get this bread no matter what"]]))                                                                                                    

(defn goodbye
   "A route for goodbye"
   [request]
   (html [:h1 "Thank you so much for your visit."]
         [:p "We are very grateful of your presence!!"] ))


(defn about 
  "It gives you info about the prog."
  [request]
  {:status 200
   :body "<h1>Nkosinathi's simple web service, not bad...</h1>"
   :headers {}})

(defn req-info
  "Gives you information about the request that has been sent"
  [request]
  
  {:status 200
   :body (pr-str request)
   :headers {}})

  (defn hello
  "Takes in a name from the request and display a personalised message"
  [request]
  (let [name (get-in request [:route-params :name])]
    {:status 200
     :body (str "Hello, "name ". How are you today ?")
     :headers {}}))

(def operators {"+" + "-" - ":" / "*" *})

(defn calculator
  "Gives a calculated result from the request's parameters"
  [request]
 (let [a  (Integer. (get-in request [:route-params :a]))
       b  (Integer. (get-in request [:route-params :b]))
       op (get-in request [:route-params :op])  
       the_op (get operators op)]
      
      
      
  (if the_op
    {:status 200
     :body  (str "Your result: " (the_op a b)) 
     :headers {}}
    {:status 404
     :body "I am not farmiliar with that operand. Try +,-,* or /"
     :headers {}})))


(defroutes app

  (GET "/" [] the-handler)
  (GET "/goodbye" [] goodbye)
  (GET "/about" [] about)
  (GET "/req-info" [] req-info)
  (GET "/hello/:name" [] hello)
  (GET "/calculator/:op/:a/:b" [] calculator)
  (not-found  "<h1>This is not the page you are looking for</h1>
            <p>Sorry, the page you requested was not found!</p>")) 

(defn -main
  "The -main function takes a port number as an argument which we pass when running the application. 
  The Ring Jetty adaptor is used to run an instance of Jetty. 
  The -main function contains an anonymous function that takes any request and returns a response map."
  [pn]  ;The port number

  (jetty/run-jetty 
   app
   {:port (Integer. pn)})
  )

(defn -dev-main
  "This one allows us to make changes and save then it pushes all those to the server."
  [pn]  ;The port number

  (jetty/run-jetty 
   (wrap-reload #'app) ; Skip the evaluation of the function and use the name instead
   {:port (Integer. pn)})
  )



